3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala

```diff
@@ -106,22 +163,23 @@ case object UaParserEnrichment extends Enrichment {
    *         exception, boxed in a
    *         Scalaz Validation
    */
-  def extractUserAgent(useragent: String): Validation[String, JsonAST.JObject] = {
-
-    val c = try {
-      uaParser.parse(useragent)
-    } catch {
-      case NonFatal(e) => return "Exception parsing useragent [%s]: [%s]".format(useragent, e.getMessage).fail
-    }
-    // To display useragent version
-    val useragentVersion = checkNull(c.userAgent.family) + prependSpace(c.userAgent.major) + prependDot(
-      c.userAgent.minor) + prependDot(c.userAgent.patch)
+  def extractUserAgent(useragent: String): Validation[String, JsonAST.JObject] =
+    for {
+      parser <- uaParser
+      c <- try {
+        parser.parse(useragent).success
+      } catch {
+        case NonFatal(e) => "Exception parsing useragent [%s]: [%s]".format(useragent, e.getMessage).fail
+      }
+    } yield {
+      // To display useragent version
```

Comment:
Can we factor out it into a separate function. I feel bad about having such big expression in `yield`.